api/app/registration/tests.py

```diff
@@ -201,10 +201,157 @@ def test_update_offer(self):
         self.assertEqual(response.status_code, 201)
         self.assertTrue(data['candidate_response'])
 
+class OfferTagAPITest(ApiTestCase):
+
+    def _seed_static_data(self):
+        test_user = self.add_user('something@email.com')
+        offer_admin = self.add_user('offer_admin@ea.com', 'event_admin', is_admin=True)
+        self.add_organisation('Deep Learning Indaba', 'blah.png', 'blah_big.png', 'deeplearningindaba')
+        db.session.add(UserCategory('Offer Category'))
+        db.session.add(Country('Suid Afrika'))
+        db.session.commit()
+
+        event = self.add_event(
+            name={'en': "Tech Talk"},
+            description={'en': "tech talking"},
+            start_date=datetime(2019, 12, 12),
+            end_date=datetime(2020, 12, 12),
+            key='SPEEDNET'
+        )
+        db.session.commit()
+
+        app_form = self.create_application_form()
+        self.add_response(app_form.id, test_user.id, True, False)
+
+        self.offer = Offer(
+            user_id=test_user.id,
+            event_id=event.id,
+            offer_date=datetime.now(),
+            expiry_date=datetime.now() + timedelta(days=15),
+            payment_required=False,
+            travel_award=True,
+            accommodation_award=False)
+        self.tag1 = self.add_tag()
+        self.tag2 = self.add_tag(names={'en': 'Tag 2 en', 'fr': 'Tag 2 fr'})
+
+        self.tag_offer(self.offer.id, self.tag1.id)
+        self.tag_offer(self.offer.id, self.tag2.id)
+
+        db.session.add(self.offer)
+        db.session.commit()
+
+        self.headers = self.get_auth_header_for("something@email.com")
+        self.adminHeaders = self.get_auth_header_for("offer_admin@ea.com")
+
+        self.add_email_template('offer')
+
+        db.session.flush()
+
+    def test_tag_admin(self):
```

Comment:
Ah because self.offer.id will only be set when it gets added to the database, which happens after db.session.add(self.offer) and db.session.commit(). But in line 238 you're trying to access self.offer.id before the offer object has been  committed to the database, so you will need to rearrange those lines to commit the offer first and then call tag_offer. I think there's also another way to do it by setting the tag objects on the offer object's tags property (so don't use the ids) and then you can commit the entire object graph in one go, but I've usually just done it step by step, create object, commit it, add other stuff, commit etc.

Reply:
Ah I see. Thought when I do shift it afterwards, I get this message instead now:
sqlalchemy.orm.exc.DetachedInstanceError: Instance <InvitedGuest at 0x7f073d967b10> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: http://sqlalche.me/e/13/bhk3)

Reply:
Oh no the dreaded DetachedInstanceError! :see_no_evil: 
This is a weird one, Amrit and I have been trying to figure it out for ages now, think we've concluded that it's some bug in SqlAlchemy or the way we're using it. Anyway, it only seems to happen sometimes, but if it does happen it is usually when you try to access an object property after doing an API call (which you're doing in some of your tests where you have a POST, then after the POST you're accessing self.offer.id and self.event1.id before doing a GET). It seems that something about the API call is causing these objects to become detached from their sessions. Anyway, the workaround which you'll see in many of our tests is to instead cache the properties that you want to use in the seed_static_data function, so for example [in this test](https://github.com/deep-learning-indaba/Baobab/blob/91b49ab42323ce480204a3bf27a27768b654d0aa/api/app/outcome/tests.py#L49) you'll see that we cache the IDs and use those instead of calling `self.event1_user1_outcome1.id`